litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,14 +1922,82 @@ protected LayoutState calculateLayoutState(
       @Nullable LayoutState previousLayoutState,
       @Nullable TreeProps treeProps,
       @CalculateLayoutSource int source,
-      String extraAttribution) {
+      @Nullable String extraAttribution) {
+
+    if (mUseSharedLayoutStateFuture) {
+      LayoutStateFuture localLayoutStateFuture = new LayoutStateFuture(
+          lock,
+          context,
+          root,
+          widthSpec,
+          heightSpec,
+          diffingEnabled,
+          previousLayoutState,
+          treeProps,
+          source,
+          extraAttribution
+      );
+
+      synchronized (mLayoutStateFutureLock) {
+        if (localLayoutStateFuture.equals(mLayoutStateFuture) && !mLayoutStateFuture.isReleased()) {
+          // Use the latest LayoutState calculation if it's the same and its LayoutState unreleased.
+          localLayoutStateFuture = mLayoutStateFuture;
+        } else {
+          // Otherwise set our calculation as the latest one.
+          mLayoutStateFuture = localLayoutStateFuture;
+        }
+      }
+      localLayoutStateFuture.run();
+
+      final LayoutState layoutState = localLayoutStateFuture.get();
+      if (layoutState == null) {
+        // If the LayoutState is already released, recalculate.
+        return calculateLayoutState(
+            lock,
+            context,
+            root,
+            widthSpec,
+            heightSpec,
+            diffingEnabled,
+            previousLayoutState,
+            treeProps,
+            source,
+            extraAttribution);
+      }
+      return layoutState;
+    } else {
+      return calculateLayoutStateInternal(
+          lock,
+          context,
+          root,
+          widthSpec,
+          heightSpec,
+          diffingEnabled,
+          previousLayoutState,
+          treeProps,
+          source,
+          extraAttribution);
+    }
+  }
+
+  private LayoutState calculateLayoutStateInternal(
+      @Nullable Object lock,
+      ComponentContext context,
+      Component root,
+      int widthSpec,
+      int heightSpec,
+      boolean diffingEnabled,
+      @Nullable LayoutState previousLayoutState,
+      @Nullable TreeProps treeProps,
+      @CalculateLayoutSource int source,
+      @Nullable String extraAttribution) {
     final ComponentContext contextWithStateHandler;
 
-    synchronized (this) {
-      final KeyHandler keyHandler =
-          (ComponentsConfiguration.useGlobalKeys || ComponentsConfiguration.isDebugModeEnabled)
-              ? new KeyHandler(mContext.getLogger())
-              : null;
+    synchronized (ComponentTree.this) {
```

Comment:
Don't need `ComponentTree.` here anymore right? Actually I think this whole block can be reverted

Reply:
done